File: script/generate_app_icon.swift
#!/usr/bin/env swift

import AppKit
import Foundation

let arguments = CommandLine.arguments
guard arguments.count == 2 else {
    fputs("Usage: generate_app_icon.swift <output.icns>\n", stderr)
    exit(64)
}

let outputURL = URL(fileURLWithPath: arguments[1])
let fileManager = FileManager.default
let iconsetURL = outputURL
    .deletingLastPathComponent()
    .appendingPathComponent("AppIcon.iconset")

try? fileManager.removeItem(at: iconsetURL)
try fileManager.createDirectory(at: iconsetURL, withIntermediateDirectories: true)

let iconSpecs: [(name: String, size: CGFloat, scale: CGFloat)] = [
    ("icon_16x16.png", 16, 1),
    ("icon_16x16@2x.png", 16, 2),
    ("icon_32x32.png", 32, 1),
    ("icon_32x32@2x.png", 32, 2),
    ("icon_128x128.png", 128, 1),
    ("icon_128x128@2x.png", 128, 2),
    ("icon_256x256.png", 256, 1),
    ("icon_256x256@2x.png", 256, 2),
    ("icon_512x512.png", 512, 1),
    ("icon_512x512@2x.png", 512, 2)
]

func roundedRectPath(in rect: NSRect, radius: CGFloat) -> NSBezierPath {
    NSBezierPath(roundedRect: rect, xRadius: radius, yRadius: radius)
}

func drawGradientBackground(in rect: NSRect) {
    let colors = [
        NSColor(calibratedRed: 0.08, green: 0.29, blue: 0.72, alpha: 1),
        NSColor(calibratedRed: 0.10, green: 0.56, blue: 0.92, alpha: 1)
    ]
    let gradient = NSGradient(colors: colors)
    let path = roundedRectPath(in: rect.insetBy(dx: rect.width * 0.05, dy: rect.height * 0.05), radius: rect.width * 0.22)
    path.addClip()
    gradient?.draw(in: rect, angle: 42)
}

func drawWaveform(in rect: NSRect) {
    let centerY = rect.midY
    let barWidth = rect.width * 0.055
    let gap = rect.width * 0.038
    let heights = [0.22, 0.42, 0.64, 0.82, 0.54, 0.34].map { rect.height * $0 }
    let totalWidth = CGFloat(heights.count) * barWidth + CGFloat(heights.count - 1) * gap
    var x = rect.midX - totalWidth / 2

    NSColor.white.withAlphaComponent(0.94).setFill()
    for height in heights {
        let barRect = NSRect(
            x: x,
            y: centerY - height / 2,
            width: barWidth,
            height: height
        )
        roundedRectPath(in: barRect, radius: barWidth / 2).fill()
        x += barWidth + gap
    }
}

func drawSwitchMark(in rect: NSRect) {
    let markRect = NSRect(
        x: rect.minX + rect.width * 0.60,
        y: rect.minY + rect.height * 0.17,
        width: rect.width * 0.22,
        height: rect.height * 0.22
    )

    NSColor.white.withAlphaComponent(0.22).setFill()
    NSBezierPath(ovalIn: markRect.insetBy(dx: -rect.width * 0.045, dy: -rect.height * 0.045)).fill()

    NSColor.white.setStroke()
    let lineWidth = max(2, rect.width * 0.028)
    let arrow = NSBezierPath()
    arrow.lineWidth = lineWidth
    arrow.lineCapStyle = .round
    arrow.lineJoinStyle = .round
    arrow.move(to: NSPoint(x: markRect.minX, y: markRect.midY))
    arrow.line(to: NSPoint(x: markRect.maxX, y: markRect.midY))
    arrow.line(to: NSPoint(x: markRect.maxX - markRect.width * 0.26, y: markRect.maxY))
    arrow.move(to: NSPoint(x: markRect.maxX, y: markRect.midY))
    arrow.line(to: NSPoint(x: markRect.maxX - markRect.width * 0.26, y: markRect.minY))
    arrow.stroke()
}

func makeIcon(size: CGFloat, scale: CGFloat) throws -> Data {
    let pixelSize = Int(size * scale)
    let image = NSImage(size: NSSize(width: pixelSize, height: pixelSize))
    image.lockFocus()

    let rect = NSRect(x: 0, y: 0, width: pixelSize, height: pixelSize)
    NSGraphicsContext.current?.imageInterpolation = .high
    drawGradientBackground(in: rect)
    drawWaveform(in: rect.insetBy(dx: rect.width * 0.16, dy: rect.height * 0.18))
    drawSwitchMark(in: rect)

    image.unlockFocus()

    guard let tiff = image.tiffRepresentation,
          let bitmap = NSBitmapImageRep(data: tiff),
          let png = bitmap.representation(using: .png, properties: [:]) else {
        throw NSError(domain: "IconGeneration", code: 1)
    }
    return png
}

for spec in iconSpecs {
    let data = try makeIcon(size: spec.size, scale: spec.scale)
    try data.write(to: iconsetURL.appendingPathComponent(spec.name))
}

let process = Process()
process.executableURL = URL(fileURLWithPath: "/usr/bin/iconutil")
process.arguments = [
    "-c", "icns",
    iconsetURL.path,
    "-o", outputURL.path
]
try process.run()
process.waitUntilExit()

guard process.terminationStatus == 0 else {
    fputs("iconutil failed with status \(process.terminationStatus)\n", stderr)
    exit(process.terminationStatus)
}

try? fileManager.removeItem(at: iconsetURL)

File: script/generate_app_icon.py
#!/usr/bin/env python3
"""生成 DoubaoVoiceSwitch 的 macOS app icon。"""

from __future__ import annotations

import os
import subprocess
from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter


ROOT = Path(__file__).resolve().parents[1]
RESOURCE_DIR = ROOT / "Sources" / "DoubaoVoiceSwitch" / "Resources"
ICONSET_DIR = RESOURCE_DIR / "AppIcon.iconset"
ICON_FILE = RESOURCE_DIR / "AppIcon.icns"
SOURCE_PNG = RESOURCE_DIR / "AppIcon-1024.png"


def draw_voice_mark(
    draw: ImageDraw.ImageDraw,
    bounds: tuple[int, int, int, int],
    color: tuple[int, int, int, int],
    background: tuple[int, int, int, int],
) -> None:
    """绘制极简语音恢复符号。"""
    x0, y0, x1, y1 = bounds
    width = x1 - x0
    height = y1 - y0
    stroke = max(16, int(width * 0.075))
    radius = int(width * 0.22)

    draw.rounded_rectangle(
        bounds,
        radius=radius,
        outline=color,
        width=stroke,
    )

    draw.rectangle(
        (
            x0 + width * 0.62,
            y0 - stroke,
            x1 + stroke,
            y0 + height * 0.36,
        ),
        fill=background,
    )

    bar_width = max(14, int(width * 0.085))
    center_y = y0 + height * 0.52
    for center_x_ratio, bar_height_ratio in [(0.38, 0.32), (0.50, 0.54), (0.62, 0.32)]:
        center_x = x0 + width * center_x_ratio
        bar_height = height * bar_height_ratio
        draw.rounded_rectangle(
            (
                center_x - bar_width / 2,
                center_y - bar_height / 2,
                center_x + bar_width / 2,
                center_y + bar_height / 2,
            ),
            radius=bar_width // 2,
            fill=color,
        )


def make_source_icon() -> Image.Image:
    """生成 1024px 源图。"""
    size = 1024
    icon = Image.new("RGBA", (size, size), (0, 0, 0, 0))

    card = (92, 92, 932, 932)
    shadow = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    shadow_draw = ImageDraw.Draw(shadow)
    shadow_draw.rounded_rectangle(card, radius=190, fill=(0, 0, 0, 75))
    icon.alpha_composite(shadow.filter(ImageFilter.GaussianBlur(30)), (0, 22))

    draw = ImageDraw.Draw(icon)
    surface = (246, 247, 250, 255)
    draw.rounded_rectangle(card, radius=190, fill=surface)
    draw.rounded_rectangle((108, 108, 916, 916), radius=174, outline=(255, 255, 255, 165), width=4)
    draw_voice_mark(
        draw,
        (270, 270, 754, 754),
        color=(46, 123, 246, 255),
        background=surface,
    )
    return icon


def write_iconset(source: Image.Image) -> None:
    """写入 iconset 所需的所有尺寸。"""
    ICONSET_DIR.mkdir(parents=True, exist_ok=True)
    sizes = [
        ("icon_16x16.png", 16),
        ("icon_16x16@2x.png", 32),
        ("icon_32x32.png", 32),
        ("icon_32x32@2x.png", 64),
        ("icon_128x128.png", 128),
        ("icon_128x128@2x.png", 256),
        ("icon_256x256.png", 256),
        ("icon_256x256@2x.png", 512),
        ("icon_512x512.png", 512),
        ("icon_512x512@2x.png", 1024),
    ]
    for filename, icon_size in sizes:
        resized = source.resize((icon_size, icon_size), Image.Resampling.LANCZOS)
        resized.save(ICONSET_DIR / filename)


def main() -> None:
    """生成 PNG、iconset 和 icns。"""
    RESOURCE_DIR.mkdir(parents=True, exist_ok=True)
    source = make_source_icon()
    source.save(SOURCE_PNG)
    write_iconset(source)
    if ICON_FILE.exists():
        ICON_FILE.unlink()
    subprocess.run(["/usr/bin/iconutil", "-c", "icns", str(ICONSET_DIR)], check=True)
    os.replace(RESOURCE_DIR / "AppIcon.icns", ICON_FILE)


if __name__ == "__main__":
    main()
